File: Python/Medium/Array/78. Subsets.py
from typing import List


def subsets(nums: List[int]) -> List[List[int]]:

    res = [[]]
    for num in nums:
        new_subs = []
        for curr in res:
            new_subs.append(curr + [num])
        res.extend(new_subs)
    return res

    # res = [[]]
    # for num in nums:
    #     new_subs = [curr + [num] for curr in res]
    #     res.extend(new_subs)
    # return res




nums = [1, 2, 3]
print(subsets(nums))

File: PHP/Medium/Array/78.Subsets.php
<?php
function subsets($nums)
{
    $result = [[]];

    foreach ($nums as $num) {
        $newSubsets = [];

        foreach ($result as $subset) {
            $newSubset = $subset;
            $newSubset[] = $num;
            $newSubsets[] = $newSubset;
        }
        $result = array_merge($result, $newSubsets);
    }
    return $result;
}


$num = [1, 2, 3];
print_r(subsets($num));